lib/promscrape/discovery/ec2/api.go

```diff
@@ -135,19 +254,79 @@ func getMetadataByPath(apiPath string) ([]byte, error) {
 	return readResponseBody(resp, apiURL)
 }
 
-func getAPIResponse(cfg *apiConfig, action, nextPageToken string) ([]byte, error) {
+// refreshRoleArnCredentials - retrieves credentials data from aws sts
+// using role_arn
+// updates apiConfig
+func (a *apiConfig) refreshRoleArnCredentials() error {
```

Comment:
Please change signature for this function to something like the following:

```go
func getRoleARNCredentials(cfg *apiConfig) (*roleCredentials, error)
```

It shouldn't change `cfg` fields - it must return `*roleCredentials`, which should have the following fields:
```go
type roleCredentials {
  accessKey string
  secretKey string
  token string
  expiration time.Time
}
```

Then the caller may properly handle the returned credentials. This should make the resulting code more clear.